src/main/java/edu/harvard/iq/dataverse/authorization/AuthenticationServiceBean.java

```diff
@@ -978,4 +976,73 @@ public ApiToken getValidApiTokenForUser(User user) {
         }
         return apiToken;
     }
+
+    /**
+     * Looks up an authenticated user based on the provided OIDC bearer token.
+     *
+     * @param bearerToken The OIDC bearer token.
+     * @return An instance of {@link AuthenticatedUser} representing the authenticated user.
+     * @throws AuthorizationException If the token is invalid or no OIDC provider is configured.
+     */
+    public AuthenticatedUser lookupUserByOIDCBearerToken(String bearerToken) throws AuthorizationException {
+        // TODO: Get the identifier from an invalidating cache to avoid lookup bursts of the same token.
+        // Tokens in the cache should be removed after some (configurable) time.
+        OIDCUserInfo oidcUserInfo = verifyOIDCBearerTokenAndGetUserIdentifier(bearerToken);
+        return lookupUser(oidcUserInfo.getUserRecordIdentifier());
+    }
+
+    /**
+     * Verifies the given OIDC bearer token and retrieves the corresponding OIDC user info.
+     *
+     * @param bearerToken The OIDC bearer token.
+     * @return An {@link OIDCUserInfo} containing the user's identifier and user info.
+     * @throws AuthorizationException If the token is invalid or if no OIDC providers are available.
+     */
+    public OIDCUserInfo verifyOIDCBearerTokenAndGetUserIdentifier(String bearerToken) throws AuthorizationException {
+        try {
+            BearerAccessToken accessToken = BearerAccessToken.parse(bearerToken);
+            List<OIDCAuthProvider> providers = getAvailableOidcProviders();
+
+            // Ensure at least one OIDC provider is configured to validate the token.
+            if (providers.isEmpty()) {
+                logger.log(Level.WARNING, "Bearer token detected, no OIDC provider configured");
+                throw new AuthorizationException(BundleUtil.getStringFromBundle("authenticationServiceBean.errors.bearerTokenDetectedNoOIDCProviderConfigured"));
+            }
+
+            // Attempt to validate the token with each configured OIDC provider.
+            for (OIDCAuthProvider provider : providers) {
+                try {
+                    // Retrieve both user identifier and user info
+                    Optional<UserRecordIdentifier> userRecordIdentifier = provider.getUserIdentifier(accessToken);
+                    Optional<UserInfo> userInfo = provider.getUserInfo(accessToken);
+
+                    // If either is present, return the result
+                    if (userRecordIdentifier.isPresent() || userInfo.isPresent()) {
+                        logger.log(Level.FINE, "Bearer token detected, provider {0} confirmed validity and provided user info", provider.getId());
+                        return new OIDCUserInfo(userRecordIdentifier.get(), userInfo.get());
```

Comment:
If you return the OAuth2UserRecord from getUserRecord as suggested above, do you have all the info you need? Or could you add the missing fields to it? Versus creating a new aggregator class?

Reply:
I have refactored the code to rely solely on OAuth2UserRecord. Good catch, thank you.